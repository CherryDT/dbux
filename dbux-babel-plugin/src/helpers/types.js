/**
 * Helpers to generate partial AST's.
 */

import { parse } from '@babel/parser';
import { codeFrameColumns } from "@babel/code-frame";
import traverse from "@babel/traverse";

/**
 * Generates AST from source through @babel/parser
 * @param {*} source 
 */
export function parseSource(source) {
  let ast;
  try {
    source = `${source}`;
    ast = parse(source);
  } catch (err) {
    const loc = err.loc;
    if (loc) {
      err.message +=
        "\n" +
        codeFrameColumns(source, {
          start: {
            line: loc.line,
            column: loc.column + 1,
          },
        });
    }
    throw err;
  }

  const nodes = ast.program.body;
  nodes.forEach(n => traverse.removeProperties(n));
  return nodes;
}

// https://stackoverflow.com/questions/35925798/how-to-add-an-import-to-the-file-with-babel