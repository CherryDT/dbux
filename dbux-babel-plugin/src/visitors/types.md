* `MemberExpression`
  * `OptionalMemberExpression`
* `Identifier`

* `AssignmentExpression`
* `VariableDeclarator`
* `VariableDeclaration`
* `ClassPrivateProperty`
* `ClassProperty`

* `Function`

* `CallExpression`
* `OptionalCallExpression`
* `NewExpression`
* `Super`

* Arithmetic
  * `BinaryExpression`
  * `UnaryExpression`
  * `LogicalExpression`

* `ConditionalExpression`
* `SequenceExpression`

* Value-creating/-changing
  * `UpdateExpression`
  * `TemplateLiteral`
  * `ObjectExpression`

* return/yield
  * `ReturnStatement`
  * `YieldExpression`

* `AwaitExpression`

* `Decorator`


# Statements

* `Declaration`
* `BreakStatement`
* `ContinueStatement`
* `ThrowStatement`
* `ForStatement`
* `ForInStatement`
* `ForOfStatement`
* `DoWhileLoop`
* `WhileStatement`
* `IfStatement`
* `SwitchStatement`
* `SwitchCase`
* `TryStatement`
* `CatchClause`
* `ExpressionStatement`
