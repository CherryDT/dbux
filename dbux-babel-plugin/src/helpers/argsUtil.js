import EmptyArray from '@dbux/common/src/util/EmptyArray';


function makeSpreadableArgumentCfg(argPath) {
  return {
    isSpread: argPath.isSpreadElement()
  };
}

/**
 * Takes a an array of arguments to indicate which are `SpreadElement` and which not.
 * 
 * NOTE: This is used by `CallExpression`, `ArrayExpression`.
 */
export function makeSpreadableArgumentArrayCfg(argumentPaths) {
  return argumentPaths?.map(makeSpreadableArgumentCfg) || EmptyArray;
}

/**
 * Takes a an array of arguments to indicate which are `SpreadElement` and which not.
 * 
 * NOTE: This is used by `ObjectExpression`.
 */
export function makeSpreadableArgumentObjectCfg(propertyPaths) {
  return propertyPaths?.map((propPath) => ({ 
    key: propPath.node.key,
    isSpread: propPath.isSpreadElement()
  })) || EmptyArray;
}
