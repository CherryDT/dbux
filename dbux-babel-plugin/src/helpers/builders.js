/**
 * Helpers to generate partial AST's.
 */

import { parse } from '@babel/parser';
import { codeFrameColumns } from "@babel/code-frame";
import traverse from "@babel/traverse";
import * as t from "@babel/types";
import { template } from '@babel/core';

export function buildNamedExport(ids) {
  return t.exportNamedDeclaration(null, ids.map(id => t.exportSpecifier(id, id)))
}


const errHandlerTemplate = template(`
console.error(err);
throw err;
`);

export function buildTryFinally(tryNodes, finallyNodes) {
  if (tryNodes.length === 1 && t.isBlockStatement(tryNodes[0])) {
    tryNodes = tryNodes[0];
  }
  else {
    tryNodes = t.blockStatement(tryNodes);
  }

  const catchClause = null;
  // const catchClause = t.catchClause(
  //   t.identifier('err'),
  //   t.blockStatement(errHandlerTemplate({
  //   }))
  // );
  return t.tryStatement(tryNodes, catchClause, t.blockStatement(finallyNodes));
}

export function buildBlock(statements) {
  return t.blockStatement(Array.isArray(statements) ? statements : [statements]);
}

export function buildVarDecl(ids) {
  const kind = "let";
  const declarations = ids.map((id, i) => {
    return t.variableDeclarator(id, null);
  });
  return t.variableDeclaration(kind, declarations);
}

export function buildVarAssignments(ids, values) {
  const assignments = ids.map((id, i) => {
    return t.expressionStatement(t.assignmentExpression('=', id, values[i]));
  });
  return assignments;
}

/**
 * Wrap a block with a try/finally pair
 */
export function buildWrapTryFinally(tryNodes, finallyBody) {
  tryNodes = Array.isArray(tryNodes) ? tryNodes : [tryNodes];
  return buildTryFinally(tryNodes, finallyBody);
}

export function buildProgram(origPathOrNode, bodyNodes) {
  const newProgramNode = t.cloneNode(origPathOrNode.node || origPathOrNode);
  newProgramNode.body = bodyNodes;
  return newProgramNode;
}

/**
 * Build AST from source through @babel/parser
 * @param {String} source 
 * @return {SourceNode[]}
 */
export function buildSource(source) {
  let ast;
  try {
    source = `${source}`;
    ast = parse(source);
  } catch (err) {
    const loc = err.loc;
    if (loc) {
      err.message +=
        "\n" +
        codeFrameColumns(source, {
          start: {
            line: loc.line,
            column: loc.column + 1,
          },
        });
    }
    throw err;
  }

  const nodes = ast.program.body;
  nodes.forEach(n => traverse.removeProperties(n));
  return nodes;
}

// https://stackoverflow.com/questions/35925798/how-to-add-an-import-to-the-file-with-babel