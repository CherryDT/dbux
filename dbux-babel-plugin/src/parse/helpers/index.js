import Declaration from './Declaration';

export {
  Declaration
};