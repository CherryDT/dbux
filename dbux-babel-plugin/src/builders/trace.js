import template from '@babel/template';
import * as t from '@babel/types';
import { newLogger } from '@dbux/common/src/log/logger';
import { getPresentableString } from '../helpers/pathHelpers';
import { bindTemplate } from '../helpers/templateUtil';

// eslint-disable-next-line no-unused-vars
const { log, debug, warn, error: logError } = newLogger('builders/trace');

const Verbose = 2;

export const buildTraceId = bindTemplate(
  '%%traceId%% = %%makeTraceId%%(%%staticTraceId%%)',

  function buildTraceId(templ, state, traceIdVar, inProgramStaticTraceId) {
    // TODO: add custom trace data
    const { ids: { aliases: {
      makeTraceId
    } } } = state;

    return templ({
      makeTraceId,
      staticTraceId: t.numericLiteral(inProgramStaticTraceId),
      traceId: traceIdVar
    });
  }
);

export const buildTraceExpression = bindTemplate(
  '%%traceExpression%%(%%expr%%, %%tid%%)',
  function buildTraceExpression(templ, path, state, traceType) {
    // TODO: add custom trace data
    // const { scope } = path;
    const { ids: { aliases: {
      traceExpression
    } } } = state;

    const tid = buildTraceId(path, state, traceType);
    const expr = path.node;
    Verbose && debug('[te]', getPresentableString(path));

    // NOTE: templates only work on `Node`, not on `NodePath`, thus they lose all path-related information.
    
    // TODO: keep `path` data etc, if necessary - `onCopy(path, newPath);`

    return templ({
      traceExpression,
      expr,
      tid
    });
  }
);

export const buildTraceNoValue = function buildTraceNoValue(templ, path, state, traceType) {
  const { ids: { dbux } } = state;
  const traceId = state.traces.addTrace(path, traceType);
  // console.warn(`traces`, state.traces);
  return templ({
    dbux,
    traceId: t.numericLiteral(traceId)
  });
}.bind(null, template('%%dbux%%.t(%%traceId%%)'));

