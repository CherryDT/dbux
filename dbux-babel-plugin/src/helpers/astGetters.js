/**
 * For reference, much of AST manipulation is based on babel-types:
 *   Source: https://github.com/babel/babel/tree/master/packages/babel-types
 *   Node classifications: https://github.com/babel/babel/tree/master/packages/babel-types/src/validators/generated/index.js
 *   Docs: https://babeljs.io/docs/en/babel-types
 *   ASTExplorer.net: https://astexplorer.net/
 */

import groupBy from 'lodash/groupBy';
import * as t from '@babel/types';
import { buildSource } from './builders';
import {isDebug} from 'dbux-common/src/util/nodeUtil';


/**
 * Might return null if given node has no ids (e.g. anonymous function declaration)
 */
export function getAllIdsOfDeclaration(node) {
  if (node.declaration.declarations) {
    // special case: variable declarations
    //  see: https://babeljs.io/docs/en/babel-types#variabledeclaration
    return node.declaration.declarations.map(n => n.id);
  }
  return node.declaration.id && [node.declaration.id] || null;
}

/**
 * 
 */
export function getAllClassParents(path) {
  const classParents = [];
  let classParent = path;
  do {
    classParent = classParent.findParent(p => t.isClass(p.node));
    if (classParent) {
      classParents.push(classParent);
    }
  } while (classParent);
  return classParents;
}