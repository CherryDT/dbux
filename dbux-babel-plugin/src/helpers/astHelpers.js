/**
 * For reference, much of AST manipulation is based on babel-types:
 *   Source: https://github.com/babel/babel/tree/master/packages/babel-types
 *   Node classifications: https://github.com/babel/babel/tree/master/packages/babel-types/src/validators/generated/index.js
 *   Docs: https://babeljs.io/docs/en/babel-types
 *   ASTExplorer.net: https://astexplorer.net/
 */

import groupBy from 'lodash/groupBy';
import * as t from '@babel/types';
import { buildSource } from './builders';


/**
 * Might return null if given node has no ids (e.g. anonymous function declaration)
 */
export function getAllIdsOfDeclaration(node) {
  if (node.declaration.declarations) {
    // special case: variable declarations
    //  see: https://babeljs.io/docs/en/babel-types#variabledeclaration
    return node.declaration.declarations.map(n => n.id);
  }
  return node.declaration.id && [node.declaration.id] || null;
}