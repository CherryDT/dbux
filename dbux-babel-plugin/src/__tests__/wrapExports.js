import { runAllSnapshotTests } from '../testing/test-util';
import { buildVarDecl, buildVarAssignments, buildWrapTryFinally } from '../helpers/builders';
import { replaceProgramBody } from '../helpers/modification';
import { logInternalError } from '../log/logger';
import * as t from '@babel/types';

const codes = [
  `
//export default class A {};
const a=1,b=2,d=3,x=4;
export { a, b };
export class B {};
export const c = 1;
export default d;
//export c from 'x';
export * from 'hi';
`
];

const visited = new Set();


function extractExportNamedVariableDeclaration(path, node, exportIds, newIds, bodyNodes, exportNodes) {
  // variable declaration
  const decls = node.declaration.declarations;
  const nodeExportIds = decls.map(n => n.id);
  exportIds.push(...nodeExportIds);
  const newNodeIds = nodeExportIds.map(id => path.scope.generateUidIdentifier(id.name));
  newIds.push(...newNodeIds);
  exportNodes.push(
    t.exportNamedDeclaration(null,
      nodeExportIds.map((exportId, i) => t.exportSpecifier(newNodeIds[i], exportId))
    )
  );
  bodyNodes.push(node.declaration);   // keep declaration in body
}

function extractExportNamedDeclaration(path, node, exportIds, newIds, bodyNodes, exportNodes) {
  if (node.declaration) {
    const { id } = node.declaration;
    if (t.isVariableDeclaration(node)) {
      extractExportNamedVariableDeclaration(path, node, exportIds, newIds, bodyNodes, exportNodes);
    }
    else if (id) {
      // non-variable declaration
      const newId = path.scope.generateUidIdentifier(id.name);
      exportIds.push(id);
      newIds.push(newId);
      exportNodes.push(
        t.exportNamedDeclaration(null,
          [t.exportSpecifier(newId, id)]
        )
      );
      bodyNodes.push(node.declaration);   // keep declaration in body
    }
    else {
      logInternalError('Cannot understand export node (named export with declaration but without id)', node.toString());
      return;
    }
  }
  else if (node.specifiers) {
    // assign local specifier to new var, then
    // export new var as exported specifier
    const nodeExportIds = node.specifiers.map(spec => spec.local);
    exportIds.push(...nodeExportIds);
    const newNodeIds = nodeExportIds.map(id => path.scope.generateUidIdentifier(id.name));
    newIds.push(...newNodeIds);
    exportNodes.push(
      t.exportNamedDeclaration(null,
        node.specifiers.map((spec, i) => t.exportSpecifier(newNodeIds[i], spec.exported))
      )
    );
  }
  else {
    logInternalError('Cannot understand export node (named export with neither declaration nor specifiers)', node.toString());
  }
}

function extractExportDefaultDeclaration(path, node, exportIds, newIds, bodyNodes, exportNodes) {
  if (node.declaration && node.declaration.id) {
    const { id } = node.declaration;
    const newId = path.scope.generateUidIdentifier(id.name);
    exportIds.push(id);
    newIds.push(newId);
    exportNodes.push(t.exportDefaultDeclaration(id));
    bodyNodes.push(node.declaration);   // keep declaration in body
  }
  else {
    exportNodes.push(node);
  }
}

function extractExportDeclaration(path, node, exportIds, newIds, bodyNodes, exportNodes) {
  if (t.isExportNamedDeclaration(node)) {
    extractExportNamedDeclaration(path, node, exportIds, newIds, bodyNodes, exportNodes);
  }
  else if (t.isExportDefaultDeclaration(node)) {
    extractExportDefaultDeclaration(path, node, exportIds, newIds, bodyNodes, exportNodes);
  }
  else if (!node.declaration) {
    exportNodes.push(node);
  }
  else {
    logInternalError('Cannot understand export node (not named, not default, but has declaration)', node.toString());
  }
}

function programVisitor(path) {
  if (visited.has(path)) return;
  visited.add(path);

  const bodyPaths = path.get('body');
  const exportIds = [];
  const newIds = [];
  const bodyNodes = [];
  const exportNodes = [];
  for (let i = 0; i < bodyPaths.length; ++i) {
    const childPath = bodyPaths[i];
    const { node } = childPath;
    if (!t.isExportDeclaration(node)) {
      bodyNodes.push(node);
    }
    else {
      extractExportDeclaration(path, node, exportIds, newIds, bodyNodes, exportNodes);
    }
  }

  // build assignment statements
  bodyNodes.push(...buildVarAssignments(newIds, exportIds));

  // finally, wrap everything up
  const newNodes = [
    buildVarDecl(newIds),
    ...buildWrapTryFinally(bodyNodes, [], []),
    ...exportNodes
  ];
  replaceProgramBody(path, newNodes);
}


const plugin = function () {
  return {
    visitor: {
      Program: programVisitor
    }
  };
};


runAllSnapshotTests(codes, __filename, plugin, true);
