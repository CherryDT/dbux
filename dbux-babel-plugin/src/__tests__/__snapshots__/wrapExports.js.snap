// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin wrapExports.es5.js: wrapExports.es5.js 1`] = `
"
/* #################################################################################### */

//export default class A {};
const a=1,b=2,d=3,x=4;
export { a, b };
export class B {};
export const c = 1;
export default d;
//export c from 'x';
export * from 'hi';

      ↓ ↓ ↓ ↓ ↓ ↓

\\"use strict\\";

exports.__esModule = true;
var _exportNames = {
  a: true,
  b: true,
  B: true
};
exports[\\"default\\"] = exports.B = exports.b = exports.a = void 0;

var _hi = require(\\"hi\\");

Object.keys(_hi).forEach(function (key) {
  if (key === \\"default\\" || key === \\"__esModule\\") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _hi[key];
});

var _a, _b, _B;

exports.B = _B;
exports.b = _b;
exports.a = _a;

try {
  /* #################################################################################### */
  //export default class A {};
  var a = 1,
      b = 2,
      _d = 3,
      x = 4;

  var B = function B() {};

  ;
  exports.a = _a = a;
  exports.b = _b = b;
  exports.B = _B = B;
} finally {}

var _default = d; //export c from 'x';

exports[\\"default\\"] = _default;
"
`;

exports[`plugin wrapExports.esNext.js: wrapExports.esNext.js 1`] = `
"
/* #################################################################################### */

//export default class A {};
const a=1,b=2,d=3,x=4;
export { a, b };
export class B {};
export const c = 1;
export default d;
//export c from 'x';
export * from 'hi';

      ↓ ↓ ↓ ↓ ↓ ↓

let _a, _b, _B;

try {
  /* #################################################################################### */
  //export default class A {};
  const a = 1,
        b = 2,
        d = 3,
        x = 4;

  class B {}

  ;
  _a = a;
  _b = b;
  _B = B;
} finally {}

export { _a as a, _b as b };
export { _B as B };
export default d; //export c from 'x';

export * from 'hi';
"
`;
